### PR TITLE
Query generator test tool

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
After a request by @onderkalaci, I created a join query generator tool which helped us test [recurring outer join PR](https://github.com/citusdata/citus/pull/6512) . Besides, it let us find a couple of bugs related to planner during release testing. (details on Issue #6636) For those reasons, I consider integrating the tool into Citus CI tests. Before doing that, I will be refactoring + improving the tool in this PR.

[Details about the tool](https://github.com/aykut-bozkurt/join_gen)

### CODE REVIEW

**Review by onderkalaci — COMMENTED** [review]
Looks very useful, lets iterate a little more before getting into main

File: src/test/regress/citus_tests/join_gen/config/config.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+interactiveMode: false
+queryCount: 10000
+queryOutFile: queries.sql
+ddlOutFile: ddls.sql
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/join_gen/config/config.yaml:4]: tweaked config params. But we can think more about that.

File: .circleci/config.yml [hunk 2]
```diff
@@ -421,6 +421,67 @@ jobs:
       - coverage:
           flags: 'test_<< parameters.pg_major >>,upgrade'
 
+  test-join-generator:
```
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply .circleci/config.yml:424]: maybe @JelteF

File: src/test/regress/citus_tests/join_gen/README.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
+## Query Generator for Postgres
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/README.md:1]: The folder structure of this new test suite is a little confusing to me. Some minor questions and suggestions:

- why is this under citus_tests? Can't we move one level up: `/src/test/regress/join_gen`
- Why do we call this as `join_gen`? Maybe make it more generic from now: `query_gen_test` or such?
- I'm fine with `join_gen/config` folder.
- However,  the existence of `join_gen/main.py` and test/citus_compare_dist_local_joins confusing to me. The latter is the main entry point for our cases, but normally one would expect the main.py to be the main entry point. So maybe:
           -  Rename main.py to something like: run_test_per_db.py 
           - Move run_test_per_db.py under "tests" folder
           - Rename "tests" folder to "bin" or such

File: src/test/regress/citus_tests/join_gen/README.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+1. Configure `interactiveMode: false`,
+2. Configure `queryCount: <total_query>`,
+3. Configure `queryOutFile: <query_file_path>` and `ddlOutFile: <ddlfile_path>`
+4. Run the command shown below
+```bash
+python main.py
+```
+
+## How to Run Citus Join Verification?
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/README.md:204]: move this part top of the README, I think that is more common

File: src/test/regress/citus_tests/join_gen/README.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+```yaml
+interactiveMode: <bool>
+```
+
+Explanation:
+```yaml
+interactiveMode: "when true, interactively prints generated ddls and queries. Otherwise, it writes them to configured files."
+```
+
+## Goal of the Tool
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/README.md:129]: This should be at the top-most part of the file

File: src/test/regress/Makefile [hunk 6]
```diff
@@ -262,6 +263,9 @@ check-arbitrary-base: all
 check-pytest:
 	pytest -n auto
 
+check-query-generator: all
```
Comment (onderkalaci) [inline src/test/regress/Makefile:266]: should we add this to `check-full`?

File: src/test/regress/citus_tests/join_gen/generator/join_gen.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
+import random
+
+from generator.random_selections import (
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/generator/join_gen.py:3]: I think a high-level code-flow explanation would be very useful for this file. It is kind of hard to follow what is going on.

File: src/test/regress/citus_tests/join_gen/main.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import signal
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/main.py:1]: as I noted, lets change the name of this file to something else. It feels like this is the entry point of the test, which is not

File: src/test/regress/citus_tests/join_gen/tests/diff-checker.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+            if not clear:
+                f.write(line)
+    return
+
+
+def removeFailedQueryOutputFromFiles(distQueryOutFile, localQueryOutFile):
+    # remove the failed distributed query from both local and distributed query files to prevent diff for acceptable errors
+    acceptableErrors = [
+        "ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns",
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/join_gen/tests/diff-checker.py:73]: Can you add a comment why are these failures happening? Is that due to https://github.com/citusdata/citus/issues/6653?

File: src/test/regress/citus_tests/join_gen/config/config.yaml [hunk 10]
```diff
@@ -0,0 +1,67 @@
+interactiveMode: false
+queryCount: 10000
+queryOutFile: queries.sql
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/citus_tests/join_gen/config/config.yaml:3]: maybe add a comment referring to this issue

File: src/test/regress/citus_tests/join_gen/config/config.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/citus_tests/join_gen/config/config.yaml:4]: can you play with the queryCount? Can you increase it as much as possible until it becomes a bottleneck in the CI? Lets be on the conservative side, as we sometime randomly generate very complex queries, which takes some time.

Also, consider this as the final change on the PR, as other changes (such as dropping EQ) might impact this

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply .circleci/config.yml]: @JelteF can you have a quick look please? Thanks!

File: src/test/regress/citus_tests/join_gen/config/config.yaml [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  - RIGHT
+  - FULL
+
+targetRteTypes:
+  - RELATION
+  - SUBQUERY
+  - CTE
+  - VALUES
+
+targetRestrictOps:
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/citus_tests/join_gen/config/config.yaml:64]: On my tests, it seems to help. Also, making VALUES less common could also help

[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/join_gen/config/config.yaml:64]: Removed from config but we still support it

File: src/test/regress/citus_tests/join_gen/README.md (hunk 3 above)
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/join_gen/README.md]: we reuse cluster creation, deletion etc.. defined in citus_tests.

File: src/test/regress/citus_tests/join_gen/generator/join_gen.py (hunk 7 above)
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/join_gen/generator/join_gen.py]: added comment block at the top of grammar syntax. should we add more?

File: src/test/regress/citus_tests/join_gen/config/config.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/join_gen/config/config.yaml]: Removed EQ restriction and VALUES rte type from config to increase non-null result rate. It worked but now test times out most of the times. I probably need to tweak more. You can suggest more on that too.

File: src/test/regress/citus_tests/query_generator/bin/citus_compare_dist_local_joins.sh [hunk 12]
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+# make bash behave
+set -euo pipefail
+
+psql_user=$1
```
Comment (onderkalaci) [inline src/test/regress/citus_tests/query_generator/bin/citus_compare_dist_local_joins.sh:6]: what a database name?

File: src/test/regress/citus_tests/join_gen/config/config.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/test/regress/citus_tests/join_gen/config/config.yaml]: Alright, I can also see that we have very little NULL values with the current settings.

It only takes 30 seconds on my local machine, but I can imagine that CI is not that fast. So, maybe we decrease the queryCount to 500? I also don't want this test to become the slowest in the CI. I'm even fine with 250 for now.

File: src/test/regress/citus_tests/query_generator/bin/citus_compare_dist_local_joins.sh (hunk 12 above)
Comment (aykut-bozkurt) [reply src/test/regress/citus_tests/query_generator/bin/citus_compare_dist_local_joins.sh:6]: Database name is chosen as same with the username by psql. This is fine for running the generator both from test suite run on test cluster and also manual run on `citus_dev` since username matches database name.

Comment (onderkalaci) [reply src/test/regress/citus_tests/query_generator/bin/citus_compare_dist_local_joins.sh:6]: But I do not always use  `citus_dev`, it is also unusual to get the username/port but not the database name. I sometimes have different databases/users.

I think we should add the database name parameter as well

File: .circleci/config.yml [hunk 13]
```diff
@@ -421,6 +421,67 @@ jobs:
       - coverage:
           flags: 'test_<< parameters.pg_major >>,upgrade'
 
+  test-query-generator:
```
Comment (onderkalaci) [inline .circleci/config.yml:424]: When I check CI, this one shows as optional. Is that intentional? I'd rather make it required.

**Review by onderkalaci — APPROVED** [review]
Please get a review for the CI things before merging (cc: @JelteF)

Please consider getting the db name param as well, that is blocking my workflow to test locally

**Comment by aykut-bozkurt** [comment]
We can remove pyyaml installation step from ci steps since I burn pyyaml package into Docker images at https://github.com/citusdata/the-process/pull/122.

File: .circleci/config.yml (hunk 13 above)
Comment (JelteF) [reply .circleci/config.yml:424]: just to be clear, if we make this required, we can only make it required after we merge this PR that adds the CI job. Otherwise all other PRs are blocked.
